File: src/test/java/integration/cardinstance/CardInstanceTest.java
package integration.cardinstance;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;

import com.matag.cards.Cards;
import com.matag.game.cardinstance.CardInstanceFactory;

import integration.TestUtils;
import integration.TestUtilsConfiguration;

@RunWith(SpringRunner.class)
@ContextConfiguration(classes = TestUtilsConfiguration.class)
public class CardInstanceTest {
  @Autowired
  private CardInstanceFactory cardInstanceFactory;

  @Autowired
  private Cards cards;

  @Autowired
  private TestUtils testUtils;

  @Rule
  public ExpectedException thrown = ExpectedException.none();

  @Test
  public void checkIfCanAttackShouldWorkForCreatures() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = cardInstanceFactory.create(gameStatus, 1, cards.get("Feral Maaka"), "player-name");

    // When
    cardInstance.checkIfCanAttack();
  }

  @Test
  public void checkIfCanAttackShouldThrowExceptionForNonCreatures() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = cardInstanceFactory.create(gameStatus, 1, cards.get("Short Sword"), "player-name");

    // Expects
    thrown.expectMessage("\"1 - Short Sword\" is not of type Creature.");

    // When
    cardInstance.checkIfCanAttack();
  }

  @Test
  public void checkIfCanAttackShouldThrowExceptionForTappedCreatures() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = cardInstanceFactory.create(gameStatus, 1, cards.get("Feral Maaka"), "player-name");
    cardInstance.getModifiers().setTapped(true);

    // Expects
    thrown.expectMessage("\"1 - Feral Maaka\" is tapped and cannot attack.");

    // When
    cardInstance.checkIfCanAttack();
  }

  @Test
  public void checkIfCanAttackShouldThrowExceptionForCreaturesWithSummoningSickness() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = cardInstanceFactory.create(gameStatus, 1, cards.get("Feral Maaka"), "player-name");
    cardInstance.getModifiers().setSummoningSickness(true);

    // Expects
    thrown.expectMessage("\"1 - Feral Maaka\" has summoning sickness and cannot attack.");

    // When
    cardInstance.checkIfCanAttack();
  }

  @Test
  public void checkIfCanAttackShouldBeOkForCreaturesWithSummoningSicknessButHaste() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = cardInstanceFactory.create(gameStatus, 1, cards.get("Nest Robber"), "player-name");
    cardInstance.getModifiers().setSummoningSickness(true);

    // When
    cardInstance.checkIfCanAttack();
  }

  @Test
  public void checkIfCanAttackShouldThrowExceptionForCreaturesWithDefender() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = cardInstanceFactory.create(gameStatus, 1, cards.get("Guardians of Meletis"), "player-name");

    // Expects
    thrown.expectMessage("\"1 - Guardians of Meletis\" has defender and cannot attack.");

    // When
    cardInstance.checkIfCanAttack();
  }
}

File: src/test/kotlin/integration/cardinstance/CardInstanceTest.kt
package integration.cardinstance

import com.matag.cards.Cards
import com.matag.game.MatagGameApplication
import com.matag.game.cardinstance.CardInstanceFactory
import com.matag.game.message.MessageException
import integration.TestUtils
import integration.TestUtilsConfiguration
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles

@SpringBootTest(classes = [MatagGameApplication::class], webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles("test")
@Import(TestUtilsConfiguration::class)
class CardInstanceTest(
    val cardInstanceFactory: CardInstanceFactory,
    val cards: Cards,
    val testUtils: TestUtils
) {
    @Test
    fun checkIfCanAttackShouldWorkForCreatures() {
        // Given
        val gameStatus = testUtils.testGameStatus()
        val cardInstance = testUtils.createCardInstance(gameStatus, "Feral Maaka")

        // When
        cardInstance.checkIfCanAttack()
    }

    @Test
    fun checkIfCanAttackShouldThrowExceptionForNonCreatures() {
        // Given
        val gameStatus = testUtils.testGameStatus();
        val cardInstance = testUtils.createCardInstance(gameStatus, "Short Sword");

        // Expects
        assertThatThrownBy { cardInstance.checkIfCanAttack() }
            .isInstanceOf(MessageException::class.java)
            .hasMessage("\"${cardInstance.id} - Short Sword\" is not of type Creature.")
    }

    @Test
    fun checkIfCanAttackShouldThrowExceptionForTappedCreatures() {
        // Given
        val gameStatus = testUtils.testGameStatus();
        val cardInstance = testUtils.createCardInstance(gameStatus, "Feral Maaka");
        cardInstance.modifiers.isTapped = true;

        // Expects
        assertThatThrownBy { cardInstance.checkIfCanAttack() }
            .isInstanceOf(MessageException::class.java)
            .hasMessage("\"${cardInstance.id} - Feral Maaka\" is tapped and cannot attack.")
    }

    @Test
    fun checkIfCanAttackShouldThrowExceptionForCreaturesWithSummoningSickness() {
        // Given
        val gameStatus = testUtils.testGameStatus();
        val cardInstance = testUtils.createCardInstance(gameStatus, "Feral Maaka");
        cardInstance.modifiers.isSummoningSickness = true;

        // Expects
        assertThatThrownBy { cardInstance.checkIfCanAttack() }
            .isInstanceOf(MessageException::class.java)
            .hasMessage("\"${cardInstance.id} - Feral Maaka\" has summoning sickness and cannot attack.")
    }

    @Test
    fun checkIfCanAttackShouldBeOkForCreaturesWithSummoningSicknessButHaste() {
        // Given
        val gameStatus = testUtils.testGameStatus()
        val cardInstance = testUtils.createCardInstance(gameStatus, "Nest Robber")
        cardInstance.modifiers.isSummoningSickness = true

        // When
        cardInstance.checkIfCanAttack()
    }

    @Test
    fun checkIfCanAttackShouldThrowExceptionForCreaturesWithDefender() {
        // Given
        val gameStatus = testUtils.testGameStatus();
        val cardInstance = testUtils.createCardInstance(gameStatus, "Guardians of Meletis");

        // Expects
        assertThatThrownBy { cardInstance.checkIfCanAttack() }
            .isInstanceOf(MessageException::class.java)
            .hasMessage("\"${cardInstance.id} - Guardians of Meletis\" has defender and cannot attack.")
    }
}